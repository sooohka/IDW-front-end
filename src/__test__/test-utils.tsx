/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable no-param-reassign */
import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import renderer from "react-test-renderer";
import store from "../store";
import GlobalStyle from "../style/GlobalStyle";
import theme from "../style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const thunk =
  ({ dispatch, getState }: any) =>
  (next: any) =>
  (action: any) => {
    if (typeof action === "function")
      // 여기서 반환되는 action은 thunk함수임
      return action(dispatch, getState);
    return next(action);
  };

const useStore = (stateName: string, initState: any, reducer: any) => {
  const store = {
    dispatch: jest.fn((action: any) => {
      initState = reducer(initState, action);
    }),
    getState: jest.fn(() => ({ [stateName]: initState })),
  };
  const next = jest.fn();
  const makeThunk = (action: any) => thunk(store)(next)(action);
  return { store, next, makeThunk };
};

const createWithThemeProvider = (EL: ReactElement) => {
  return renderer.create(
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {EL}
    </ThemeProvider>,
  );
};

const redux = { thunk, useStore };
export { redux, customRender as render, createWithThemeProvider };

export * from "@testing-library/react";
