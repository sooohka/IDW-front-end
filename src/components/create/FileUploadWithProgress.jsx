import PropTypes from "prop-types";
import React, { memo, useEffect, useState } from "react";
import styled from "styled-components";
import HelperText from "../common/HelperText";
import Img from "../common/Img";
import ProgressBar from "../common/ProgressBar";
import { ReactComponent as Spinner } from "../../assets/icons/spinner.svg";
import { ReactComponent as FileImage } from "../../assets/icons/file-image-regular.svg";
import XButton from "../common/XButton";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 1rem 1rem;
`;

const ProgressWrapper = styled.div`
  display: flex;
  width: 100%;
  align-items: center;
`;

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 0 0 0 1rem;
  width: 100%;
`;

const FileUploadWithProgress = ({ handleDelete, handleUpload, file }) => {
  const [progress, setProgress] = useState(0);
  const {
    file: { name, lastModified, lastModifiedDate, path, size, type },
    isSubmitted,
    id,
    url,
    fullUrl,
    error: { status: errorStatus, message },
  } = file;

  useEffect(() => {
    async function upload() {
      const res = await handleUpload(file, setProgress);
    }
    upload();
  }, [handleUpload]);
  console.log(isSubmitted);

  return (
    <Container onClick={(e) => e.preventDefault()}>
      {isSubmitted ? <Img width="50px" height="50px" src={fullUrl} alt={name} /> : <FileImage width={50} height={50} />}
      <Wrapper>
        <ProgressWrapper>
          <ProgressBar title={name} hasError={errorStatus} progress={progress} />
          {isSubmitted ? <XButton onClick={handleDelete(id)} /> : <Spinner />}
        </ProgressWrapper>
        <HelperText always hasError={errorStatus} text={message} />
      </Wrapper>
    </Container>
  );
  // <Template file={file} progress={progress} handleDelete={handleDelete(file.id)} />;
};

FileUploadWithProgress.propTypes = {
  handleUpload: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
  file: PropTypes.shape({
    file: PropTypes.shape({
      name: PropTypes.string.isRequired,
      lastModified: PropTypes.number.isRequired,
      lastModifiedDate: PropTypes.instanceOf(Date).isRequired,
      path: PropTypes.string.isRequired,
      size: PropTypes.number.isRequired,
      type: PropTypes.string.isRequired,
    }),
    isSubmitted: PropTypes.bool.isRequired,
    id: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    fullUrl: PropTypes.string.isRequired,
    error: PropTypes.shape({ status: PropTypes.bool.isRequired, message: PropTypes.string.isRequired }).isRequired,
  }).isRequired,
};
export default FileUploadWithProgress;
