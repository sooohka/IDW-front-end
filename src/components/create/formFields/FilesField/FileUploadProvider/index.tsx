import React, { createContext, useCallback, useContext, useReducer } from "react";

// actions
const DELETE_FILE = "file/deleteFile" as const;
const APPEND_FILES = "file/appendFiles" as const;
const UPDATE_FILE = "file/updateFile" as const;
const SET_ACCEPTING = "file/setIsAccepting" as const;

// action creators
const appendFiles = (files: TargetFile[]) => ({ type: APPEND_FILES, payload: { files } });
const updateFiles = (file: TargetFile) => ({ type: UPDATE_FILE, payload: { file } });
const deleteFile = (id: string) => ({ type: DELETE_FILE, payload: { id } });
const setIsAccepting = (isAccepting: boolean) => ({
  type: SET_ACCEPTING,
  payload: { isAccepting },
});

type Action =
  | ReturnType<typeof appendFiles>
  | ReturnType<typeof updateFiles>
  | ReturnType<typeof deleteFile>
  | ReturnType<typeof setIsAccepting>;

const initState: { files: TargetFile[]; isAccepting: boolean } = { files: [], isAccepting: false };
const reducer = (state = initState, action: Action): typeof initState => {
  switch (action.type) {
    case APPEND_FILES: {
      const { files } = action.payload;
      return { ...state, files: [...state.files, ...files] };
    }
    case UPDATE_FILE: {
      const { file } = action.payload;
      return { ...state, files: state.files.map((v) => (v.id === file.id ? file : v)) };
    }
    case DELETE_FILE: {
      const { id } = action.payload;
      return { ...state, files: state.files.filter((file) => file.id !== id) };
    }
    case SET_ACCEPTING: {
      const { isAccepting } = action.payload;
      return { ...state, isAccepting };
    }
    default:
      return state;
  }
};

interface FileUploadContextType {
  isAccepting: boolean;
  files: TargetFile[];
  handleFilesAppend: (files: TargetFile[]) => void;
  handleFileDelete: (id: string) => void;
  handleFileUpdate: (fils: TargetFile) => void;
  handleSetIsAccepting: (isAccepting: boolean) => void;
}

const FileUploadContext = createContext<FileUploadContextType>({
  isAccepting: false,
  files: [],
  handleFileDelete: () => {},
  handleFileUpdate: () => {},
  handleFilesAppend: () => {},
  handleSetIsAccepting: () => {},
});

const FileUploadProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initState);

  const handleFilesAppend = useCallback((files: TargetFile[]) => {
    dispatch(appendFiles(files));
  }, []);
  const handleFileDelete = useCallback((id: string) => {
    dispatch(deleteFile(id));
  }, []);
  const handleSetIsAccepting = useCallback((isAccepting: boolean) => {
    dispatch(setIsAccepting(isAccepting));
  }, []);
  const handleFileUpdate = useCallback((file: TargetFile) => {
    dispatch(updateFiles(file));
  }, []);

  const value = {
    files: state.files,
    isAccepting: state.isAccepting,
    handleFileDelete,
    handleFilesAppend,
    handleSetIsAccepting,
    handleFileUpdate,
  };
  return <FileUploadContext.Provider value={value}>{children}</FileUploadContext.Provider>;
};

const useFileUploadContext = () => useContext(FileUploadContext);

export { useFileUploadContext };
export default FileUploadProvider;
