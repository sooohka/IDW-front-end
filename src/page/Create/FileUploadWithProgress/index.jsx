import axios from "axios";
import React, { useCallback, useEffect, useState } from "react";
import PropTypes from "prop-types";
import Template from "./template";
import { uploadFile } from "../../../api/aws";

const FileUploadWithProgress = ({ handleDelete, handleSubmittedFiles, file }) => {
  const [progress, setProgress] = useState(0);
  const [fileInfo, setFileInfo] = useState({
    file: {
      name: file.name,
      url: "",
    },
    isSubmitting: true,
    hasError: false,
    message: "submitting...",
  });

  const handleUpload = useCallback(
    async (_file) => {
      try {
        let response;
        // aws ì‚¬ìš©
        if (process.env.REACT_APP_ENV !== "local") response = await uploadFile(_file, setProgress);
        else alert("env=devìž…ë‹ˆë‹¤.");

        const { url, fullUrl, name } = response.data;
        console.log(response.data);

        handleSubmittedFiles({ name, url });
        setFileInfo((prev) => ({ ...prev, isSubmitting: false, message: "submitted!", file: { ...prev.file, url: fullUrl } }));
      } catch (err) {
        console.error(err.message);
        // TODO: fileUploadWithProgress error handling
        const message = err.response?.data?.error?.message || "something went wrongðŸ˜… ";
        setFileInfo((prev) => ({ isSubmitting: false, hasError: true, message, file: { ...prev.file } }));
      }
    },
    [handleSubmittedFiles]
  );

  useEffect(() => {
    async function upload() {
      await handleUpload(file);
    }
    upload();
  }, [file, handleUpload]);

  return <Template type={file.type} fileInfo={fileInfo} progress={progress} handleDelete={handleDelete} />;
};

FileUploadWithProgress.propTypes = {
  handleDelete: PropTypes.func.isRequired,
  handleSubmittedFiles: PropTypes.func.isRequired,
  file: PropTypes.instanceOf(File).isRequired,
};
export default FileUploadWithProgress;
