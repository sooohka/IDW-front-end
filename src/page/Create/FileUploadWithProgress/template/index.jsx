import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import ProgressBar from "../../../../components/common/ProgressBar";
import XButton from "../../../../components/common/XButton";
import { ReactComponent as Spinner } from "../../../../assets/icons/spinner.svg";
import { ReactComponent as FileImage } from "../../../../assets/icons/file-image-regular.svg";
import HelperText from "../../../../components/common/HelperText";
import Img from "../../../../components/common/Img";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 1rem 1rem;
`;

const ProgressWrapper = styled.div`
  display: flex;
  width: 100%;
  align-items: center;
`;
const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 0 0 0 1rem;
  width: 100%;
`;
const Template = ({ isSubmitting, fileInfo, progress, handleDelete }) => {
  const {
    file: { name, lastModified, lastModifiedDate, path, size, type },
    hasError,
    message,
    url,
  } = fileInfo;

  return (
    <Container onClick={(e) => e.preventDefault()}>
      {isSubmitting ? <FileImage width={50} height={50} /> : <Img width="50px" height="50px" src={url} alt={name} />}
      <Wrapper>
        <ProgressWrapper>
          <ProgressBar title={name} hasError={hasError} progress={progress} />
          {isSubmitting ? <Spinner /> : <XButton onClick={handleDelete(name)} />}
        </ProgressWrapper>
        <HelperText always hasError={hasError} text={message} />
      </Wrapper>
    </Container>
  );
};

Template.propTypes = {
  isSubmitting: PropTypes.bool.isRequired,
  fileInfo: PropTypes.shape({
    file: PropTypes.shape({
      name: PropTypes.string.isRequired,
      lastModified: PropTypes.number.isRequired,
      lastModifiedDate: PropTypes.instanceOf(Date).isRequired,
      path: PropTypes.string.isRequired,
      size: PropTypes.number.isRequired,
      type: PropTypes.string.isRequired,
    }).isRequired,
    url: PropTypes.string.isRequired,
    hasError: PropTypes.bool.isRequired,
    message: PropTypes.string.isRequired,
  }).isRequired,
  progress: PropTypes.number.isRequired,
  handleDelete: PropTypes.func.isRequired,
};
export default Template;
