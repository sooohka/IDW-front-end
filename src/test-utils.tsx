import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import store from "./store";
import GlobalStyle from "./style/GlobalStyle";
import theme from "./style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
