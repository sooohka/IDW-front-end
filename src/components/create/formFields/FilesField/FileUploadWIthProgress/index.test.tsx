import { act, screen, render, createWithThemeProvider } from "../../../../../test/test-utils";
import FileUploadWithProgress from ".";
import { FileUploadContext, FileUploadContextType } from "../FileUploadProvider";

describe("src/components/create/formFields/FilesField/FileUploadWIthProgress", () => {
  let fileUploadWithProgressProps: { imageFile: TargetFile; setIsFileUploading: () => void } = {
    imageFile: { errors: [], file: new File([], ""), id: "", isSubmitted: false, url: "" },
    setIsFileUploading: () => {},
  };

  let fileUploadContextValue: FileUploadContextType = {
    isAccepting: false,
    files: [],
    handleFileDelete: () => {},
    handleFileUpdate: () => {},
    handleFilesAppend: () => {},
    handleSetIsAccepting: () => {},
  };
  beforeEach(() => {
    fileUploadContextValue = {
      isAccepting: false,
      files: [],
      handleFileDelete: jest.fn(),
      handleFileUpdate: jest.fn(),
      handleFilesAppend: jest.fn(),
      handleSetIsAccepting: jest.fn(),
    };

    fileUploadWithProgressProps = {
      imageFile: {
        errors: [],
        file: new File(["some"], "file.pdf", { type: "application/pdf" }),
        id: "1",
        isSubmitted: false,
        url: "",
      },
      setIsFileUploading: jest.fn(),
    };
  });

  it("match snapshot", () => {
    const component = createWithThemeProvider(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadWithProgress {...fileUploadWithProgressProps} />
      </FileUploadContext.Provider>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("calls handleFileUpdate when file is uploaded", () => {
    fileUploadWithProgressProps.imageFile = {
      errors: [],
      file: new File(["blooooob"], "file.png", { type: "image/png" }),
      id: "1",
      isSubmitted: false,
      url: "",
    };
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadWithProgress {...fileUploadWithProgressProps} />
      </FileUploadContext.Provider>,
    );

    expect(fileUploadWithProgressProps.setIsFileUploading).toBeCalledTimes(2);
    expect(fileUploadContextValue.handleFileUpdate).toBeCalledTimes(1);
    expect(fileUploadContextValue.handleFileUpdate).toBeCalledWith(
      fileUploadWithProgressProps.imageFile,
    );
  });
  it("doesn't calls uploadFile when file is already submitted", () => {
    fileUploadWithProgressProps.imageFile = {
      errors: [],
      file: new File(["blooooob"], "file.png", { type: "image/png" }),
      id: "1",
      isSubmitted: true,
      url: "",
    };
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadWithProgress {...fileUploadWithProgressProps} />
      </FileUploadContext.Provider>,
    );

    expect(fileUploadContextValue.handleFileUpdate).not.toBeCalled();
  });
  it("doesn't calls uploadFile when file has some error", () => {
    fileUploadWithProgressProps.imageFile = {
      errors: [{ code: "1", message: "has error" }],
      file: new File(["blooooob"], "file.png", { type: "image/png" }),
      id: "1",
      isSubmitted: false,
      url: "",
    };
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadWithProgress {...fileUploadWithProgressProps} />
      </FileUploadContext.Provider>,
    );

    expect(fileUploadContextValue.handleFileUpdate).not.toBeCalled();
  });
  it("renders emptyFile svg when has error", () => {
    fileUploadWithProgressProps.imageFile = {
      errors: [],
      file: new File(["blooooob"], "file.png", { type: "image/png" }),
      id: "1",
      isSubmitted: true,
      url: "",
    };
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadWithProgress {...fileUploadWithProgressProps} />
      </FileUploadContext.Provider>,
    );
    const svgEl = screen.getByTestId(/empty-file-svg/i);
    expect(svgEl).toBeInTheDocument();
  });

  it("renders x-button when has error", () => {
    fileUploadWithProgressProps.imageFile = {
      errors: [{ code: "", message: "error" }],
      file: new File(["blooooob"], "file.png", { type: "image/png" }),
      id: "1",
      isSubmitted: true,
      url: "",
    };
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadWithProgress {...fileUploadWithProgressProps} />
      </FileUploadContext.Provider>,
    );
    const xButton = screen.getByRole("button");
    expect(xButton).toBeInTheDocument();
  });

  it("renders helper text when has error", () => {
    fileUploadWithProgressProps.imageFile = {
      errors: [{ code: "", message: "someThing when wrong" }],
      file: new File(["blooooob"], "file.png", { type: "image/png" }),
      id: "1",
      isSubmitted: true,
      url: "",
    };
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadWithProgress {...fileUploadWithProgressProps} />
      </FileUploadContext.Provider>,
    );
    const helperTextEl = screen.getByRole("note");
    expect(helperTextEl).toBeInTheDocument();
    expect(helperTextEl).toHaveTextContent(/someThing when wrong/i);
  });

  it("render spinner when has no Error and is uploading", () => {
    fileUploadWithProgressProps.imageFile = {
      errors: [],
      file: new File(["blooooob"], "file.png", { type: "image/png" }),
      id: "1",
      isSubmitted: true,
      url: "",
    };
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadWithProgress {...fileUploadWithProgressProps} />
      </FileUploadContext.Provider>,
    );

    const spinnerEl = screen.getByRole("status");
    expect(spinnerEl).toBeInTheDocument();
  });
});
