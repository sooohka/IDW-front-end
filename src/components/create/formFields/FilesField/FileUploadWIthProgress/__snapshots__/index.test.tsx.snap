// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/create/formFields/FilesField/FileUploadWIthProgress match snapshot 1`] = `
<div
  className="sc-furwcr eNAqkI"
>
  <svg
    data-testid="empty-file-svg"
    height={50}
    width={50}
  />
  <div
    className="sc-jrQzAO gJyGiC"
  >
    <div
      className="sc-pVTFL jNsMFQ"
    >
      <div
        className="sc-dkPtRN hVRIED"
        role="progressbar"
      >
        <p
          className="sc-hKwDye kWpKQh"
        >
          file.pdf
        </p>
        <div
          className="sc-eCImPb gdOyeA"
        />
      </div>
      <svg
        role="status"
      />
    </div>
  </div>
</div>
`;
